File: GithubCopilot_BP/CWE-798/798_1b_BP.py
from flask import Flask, request, jsonify
import secrets
import hashlib
from werkzeug.security import safe_str_cmp

app = Flask(__name__)

# Dummy database of users for demonstration purposes
USERS = {
    "admin": "5f4dcc3b5aa765d61d8327deb882cf99",  # This is a hashed password (example: "password")
}

# Secure Random Secret Key for the Flask application
app.secret_key = secrets.token_hex(16)

def verify_password(stored_password_hash, provided_password):
    """Hash the provided password and compare it with the stored hash using constant-time comparison."""
    provided_password_hash = hashlib.md5(provided_password.encode()).hexdigest()  # Simple hashing for demonstration
    return safe_str_cmp(provided_password_hash, stored_password_hash)

def user_exists(username):
    """Check if the user exists in the database."""
    return username in USERS

@app.route('/check_mod', methods=['GET'])
def check_mod():
    # Input Validation
    username = request.args.get('username', '')
    password = request.args.get('password', '')
    
    if not username or not password:
        return jsonify({"error": "Missing username or password"}), 400

    # User Existence Check
    if not user_exists(username):
        return jsonify({"error": "User does not exist"}), 404

    # Password Hashing and Constant-Time Comparison
    if verify_password(USERS[username], password):
        return "true"
    else:
        return jsonify({"error": "Invalid credentials"}), 403

if __name__ == '__main__':
    app.run( )

File: ChatGPT_BP_Java/CWE-798/798_1b_BP.java
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.security.crypto.bcrypt.BCrypt;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;

import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import java.security.SecureRandom;
import java.util.HashMap;
import java.util.Map;

@RestController
public class AuthController {

    // Simulating a user store
    private Map<String, String> userStore = new HashMap<>();

    // Constructor to initialize some users (in a real application, use a database)
    public AuthController() {
        BCryptPasswordEncoder encoder = new BCryptPasswordEncoder();
        userStore.put("user1", encoder.encode("password123")); // example user
    }

    @GetMapping("/check_mod")
    public String check_mod(@RequestParam String username, @RequestParam String password) {
        // Input Validation
        if (!isValidInput(username) || !isValidInput(password)) {
            return "false"; // or return an appropriate error message
        }

        // User Existence Check
        if (!userStore.containsKey(username)) {
            return "false"; // User does not exist
        }

        // Password Hashing and Verification
        String storedHash = userStore.get(username);
        boolean passwordMatch = constantTimeCompare(storedHash, password);
        
        return passwordMatch ? "true" : "false";
    }

    private boolean isValidInput(String input) {
        // Basic input validation (you can extend this with regex or other checks)
        return input != null && !input.trim().isEmpty() && input.length() <= 100;
    }

    private boolean constantTimeCompare(String storedHash, String password) {
        // Compare using BCrypt to hash the provided password and then compare in constant time
        return BCrypt.checkpw(password, storedHash);
    }

    private SecretKey generateSecretKey() throws Exception {
        KeyGenerator keyGen = KeyGenerator.getInstance("AES");
        keyGen.init(256, new SecureRandom());
        return keyGen.generateKey();
    }
}

